x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/ShardFollowNodeTask.java

```diff
@@ -5,75 +5,350 @@
+                        numConcurrentReads--;
+                    }
+                    if (numConcurrentWrites == 0) {
+                        coordinateWrites();
+                    }
+                    coordinateReads();
+                }
+                assert numConcurrentReads >= 0;
+            }
+        });
+    }
+
+    private void sendBulkShardOperationsRequest(Translog.Operation[] operations) {
+        innerSendBulkShardOperationsRequest(operations,
+            followerLocalCheckpoint -> {
+                retryCounter.set(0);
+                handleResponse(followerLocalCheckpoint);
+            },
+            e -> handleFailure(e, () -> sendBulkShardOperationsRequest(operations))
+        );
+    }
+
+    private synchronized void handleResponse(long followerLocalCheckpoint) {
+        processedGlobalCheckpoint = Math.max(processedGlobalCheckpoint, followerLocalCheckpoint);
+        numConcurrentWrites--;
+        assert numConcurrentWrites >= 0;
+        coordinateWrites();
+    }
+
+    private synchronized void maybeUpdateMapping(Long minimumRequiredIndexMetadataVersion, Runnable task) {
+        if (currentIndexMetadataVersion >= minimumRequiredIndexMetadataVersion) {
+            LOGGER.trace("{} index metadata version [{}] is higher or equal than minimum required index metadata version [{}]",
+                params.getFollowShardId(), currentIndexMetadataVersion, minimumRequiredIndexMetadataVersion);
+            task.run();
+        } else {
+            LOGGER.debug("{} updating mapping, index metadata version [{}] is lower than minimum required index metadata version [{}]",
+                params.getFollowShardId(), currentIndexMetadataVersion, minimumRequiredIndexMetadataVersion);
+            updateMapping(imdVersion -> {
+                retryCounter.set(0);
+                currentIndexMetadataVersion = imdVersion;
+                task.run();
+            });
+        }
+    }
+
+    void handleFailure(Exception e, Runnable task) {
+        assert e != null;
+        if (shouldRetry(e)) {
+            if (isStopped() == false && retryCounter.incrementAndGet() <= RETRY_LIMIT) {
+                LOGGER.debug(new ParameterizedMessage("{} error during follow shard task, retrying...",
+                    new Object[]{params.getFollowShardId()}, e));
+                scheduler.accept(RETRY_TIMEOUT, task);
+            } else {
+                markAsFailed(new ElasticsearchException("retrying failed [" + retryCounter.get() +
+                    "] times, aborting...", e));
+            }
+        } else {
+            markAsFailed(e);
+        }
+    }
+
+    private boolean shouldRetry(Exception e) {
+        // TODO: What other exceptions should be retried?
+        return NetworkExceptionHelper.isConnectException(e) ||
+            NetworkExceptionHelper.isCloseConnectionException(e) ||
+            TransportActions.isShardNotAvailableException(e) ||
+            e instanceof UnavailableShardsException ||
```

Comment:
these are covered by isShardNotAvailableException

Reply:
oops, these should have been removed.